### PR TITLE
allow extension to provide custom icons to data explorer tree

### LINKED ISSUE #16374: Ability to customize icons for the nodes in the CONNECTIONS tree
**Is your feature request related to a problem? Please describe.**
We would like to populate the CONNECTIONS tree with non-sql-related nodes.

**Describe the solution or feature you'd like**
From our extension, customize the icons for a given node type (ie provide a link to the svg file in the program).

**Describe alternatives you've considered**
We currently have to choose a node type that provides an icon that is similar to what we like.

**Additional context**
This is a P0 for us.

### PR DESCRIPTION
**Pull request by alanrenmsft**
This PR fixes #16374 

This custom icon path is already supported inside the core, I am adding it to the azdata API so that it can be consumed by the extension authors, also updated the sample code. tested with both default server tree and async server tree.

image

### CODE REVIEW

File: samples/sqlservices/src/featureProviders/objectExplorerProvider.ts [hunk 1]
```diff
@@ -60,13 +63,21 @@ export class ObjectExplorerProvider implements azdata.ObjectExplorerProvider {
 				nodes: [
 					{
 						nodePath: 'root/1',
-						nodeType: 'Database',
-						label: 'abc1',
+						nodeType: '',
```
Comment (alanrenmsft) [inline samples/sqlservices/src/featureProviders/objectExplorerProvider.ts:66]: the definition of NodeInfo class needs to be reviewed, it should not be required now, I'll clean it up and add comments to azdata.d.ts later while working with the partner team. this PR is only for adding custom icon support.

File: src/sql/workbench/services/objectExplorer/browser/iconRenderer.ts [hunk 2]
```diff
@@ -22,22 +23,32 @@ class IconRenderer {
...
+			light = path.light;
+			dark = path.dark;
 		}
+		return {
+			light: this.toUri(light),
+			dark: this.toUri(dark)
+		};
+	}
+
+	private toUri(path: string | URI): URI {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/iconRenderer.ts:47]: Couple of these functions could be moved out of the class - no reason for them to be in here.

File: src/sql/azdata.proposed.d.ts [hunk 3]
```diff
@@ -921,4 +921,12 @@ declare module 'azdata' {
 		 */
 		physicalMemoryInMb?: number;
 	}
+
+	export interface NodeInfo {
+		/**
+		 * Path to the icon for the node. This property is optional.
+		 * Use this property if none of the icons defined in {@link SqlThemeIcon} fits the node.
+		 */
+		iconPath?: IconPath;
```
Comment (Charles-Gagnon) [inline src/sql/azdata.proposed.d.ts:930]: I'm not sure I like having two separate properties for the icon. Could we at least make a general `icon` property that can take a string, SqlThemeIcon or IconPath and then when we promote that up to stable we can deprecate the current iconType property?

Comment (alanrenmsft) [reply src/sql/azdata.proposed.d.ts:930]: good idea, I will make this property `icon` and add SqlThemeIcon as an optional type. currently iconType also has string as a possible type, but I don't like it, that means extensions needs to know the internal ADS class names. for the new property, the string type will be for the path of the icon.

**Review by Charles-Gagnon — APPROVED** [review]

File: src/sql/workbench/services/objectExplorer/browser/iconRenderer.ts (hunk 2 above)
Comment (alanrenmsft) [reply src/sql/workbench/services/objectExplorer/browser/iconRenderer.ts:47]: these methods are local helper methods marked as private, I don't think it makes sense to move them to anywhere else. I'd leave these here.

**Pull request merged**